### PR TITLE
dialects: (Transform) Add new selecting operations

### PR DESCRIPTION
**Pull request by JonasCrols**
In this PR, 13 new core operations get added to the transform dialect. All of them useful to find and select the correct operation in a transformation. They are:
GetConsumerOfResult, GetDefiningOp, GetParentOp, GetProducerOfOperand, GetResultOp, GetTypeOp, IncludeOp, MatchOperationEmptyOp, MatchOperationNameOp, MatchParamCmpIOp, MergeHandlesOp, ParamConstantOp, SplitHandleOp,

All operations are tested using filecheck and pytest

### CODE REVIEW

**Comment by JonasCrols** [comment]
@jorendumoulin @JosseVanDelm

File: tests/dialects/test_transform.py [hunk 1]
```diff
@@ -79,7 +79,223 @@ def test_tileop_init():
...
+        None,
+    )
+    assert isinstance(op.results[0], SSAValue)
+
+
+def test_match_empty():
+    handle = test.TestOp(result_types=[transform.AnyOpType()]).results[0]
+    op = transform.MatchOperationEmptyOp(operand_handle=handle)
+    printer = Printer()
+    printer.print(op)
```
Comment (superlopuh) [inline tests/dialects/test_transform.py:202]: I missed this before, why are these added? We should never be printing in tests

File: tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir [hunk 2]
```diff
@@ -1,6 +1,10 @@
 // RUN: xdsl-opt %s | xdsl-opt | mlir-opt --allow-unregistered-dialect | filecheck %s
 
 builtin.module attributes  {"transform.with_named_sequence"} {
+  "transform.named_sequence"() <{arg_attrs = [{transform.readonly}], function_type = (!transform.any_op) -> (), sym_name = "foo"}> ({
+  ^bb0(%arg0: !transform.any_op):
+    "transform.yield"() : () -> ()
+  }) : () -> ()
```
Comment (superlopuh) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir:7]: Please use custom syntax in tests when possible, it makes them more readable

File: xdsl/dialects/transform.py [hunk 3]
```diff
@@ -335,10 +743,27 @@ class CastOp(IRDLOperation):
...
+    def __init__(self, input: SSAValue):
+        super().__init__(operands=[input], result_types=[AnyOpType()])
+
 
 Transform = Dialect(
     "transform",
     [
+        GetConsumerOfResult,
+        GetDefiningOp,
+        # GetOperandOp,
```
Comment (superlopuh) [inline xdsl/dialects/transform.py:755]: commented code

File: xdsl/dialects/transform.py [hunk 4]
```diff
@@ -129,6 +136,367 @@ class FailurePropagationModeAttr(
...
+#     https://mlir.llvm.org/docs/Dialects/Transform/#transformget_operand-transformgetoperandop
+#     """
+
+#     name = "transform.get_operand"
+
+#     raw_position_list = prop_def(DenseArrayBase)
+#     is_inverted = prop_def(UnitAttr)
+#     is_all = prop_def(UnitAttr)
+#     target = operand_def(TransformHandleType)
+#     result = result_def(TransformHandleType)
```
Comment (superlopuh) [inline xdsl/dialects/transform.py:190]: commented code

File: xdsl/dialects/transform.py [hunk 5]
```diff
@@ -129,6 +136,352 @@ class FailurePropagationModeAttr(
...
+
+@irdl_op_definition
+class MatchParamCmpIOp(IRDLOperation):
+    """
+    https://mlir.llvm.org/docs/Dialects/Transform/#transformmatchparamcmpi-transformmatchparamcmpiop
+    """
+
+    name = "transform.match.param.cmpi"
+
+    predicate = prop_def(AnyIntegerAttr)  # Only 0, 1, 2, 3, 4, 5 are valid
```
Comment (JosseVanDelm) [inline xdsl/dialects/transform.py:382]: Maybe instead refer to `xdsl/xdsl/dialects/arith.py` for all valid ones?

File: xdsl/dialects/transform.py [hunk 6]
```diff
@@ -207,21 +560,21 @@ def __init__(
...
         dynamic_sizes: Sequence[SSAValue],
-        static_sizes: DenseArrayBase | list[int] | None = None,
+        static_sizes: DenseArrayBase | list[int],
         interchange: DenseArrayBase | list[int] | None = None,
         scalable_sizes: DenseArrayBase | list[int] | None = None,
     ):
         if isinstance(static_sizes, list):
             static_sizes = DenseArrayBase.create_dense_int_or_index(
-                IndexType(), static_sizes
+                IntegerType(64), static_sizes
```
Comment (JosseVanDelm) [inline xdsl/dialects/transform.py:569]: Why is this change necessary?

File: tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir [hunk 7]
```diff
@@ -1,6 +1,10 @@
 // RUN: xdsl-opt %s | xdsl-opt | mlir-opt --allow-unregistered-dialect | filecheck %s
```
Comment (JosseVanDelm) [inline tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir:1]: why is `--allow-unregistered-dialect` necessary here?

**Review by JosseVanDelm — APPROVED** [review]
I don't have time for a super thorough review, but this looks very nice! @jorendumoulin ?

File: xdsl/dialects/transform.py (hunk 6 above)
Comment (JonasCrols) [reply xdsl/dialects/transform.py:569]: i64 is the correct format, I found. Index are not allowed. I didn't catch it at first as I hadn't initiated an instance using int's at first

File: tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir (hunk 7 above)
Comment (JonasCrols) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir:1]: all other tests used it, so I used it too... but indeed it is not strictly necessary here

File: xdsl/dialects/transform.py [hunk 8]
```diff
@@ -129,6 +136,352 @@ class FailurePropagationModeAttr(
...
+
+    name = "transform.get_parent_op"
+
+    isolated_from_above = opt_prop_def(UnitAttr)
+    allow_empty_results = opt_prop_def(UnitAttr)
+    op_name = opt_prop_def(StringAttr)
+    deduplicate = opt_prop_def(UnitAttr)
+    nth_parent = prop_def(AnyIntegerAttr)
+    target = operand_def(TransformHandleType)
+    parent_result = result_def(TransformHandleType)
```
[the remark this answers is not in the document]
Comment (jorendumoulin) [reply xdsl/dialects/transform.py:192]: or does this overwrite something?

**Review by jorendumoulin — COMMENTED** [review]
this is looking good!

File: xdsl/dialects/transform.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (superlopuh) [reply xdsl/dialects/transform.py:192]: Unfortunately :( It overrides `parent` on IRNode, which should really be `parent_node`

File: tests/dialects/test_transform.py [hunk 9]
```diff
@@ -76,10 +75,198 @@ def test_tileop_init():
...
+    target = test.TestOp(result_types=[transform.AnyOpType()]).results[0]
+    op = transform.GetParentOp(target=target)
+    expected = """
+%0 = "transform.get_parent_op"(%1) <{"nth_parent" = 1 : i64}> : (!transform.any_op) -> !transform.any_op    """
+    assert_print_op(
+        op,
+        expected,
+        None,
+    )
+    assert isinstance(op.results[0], SSAValue)
```
Comment (superlopuh) [inline tests/dialects/test_transform.py:118]: this will always pass, let's remove these checks

File: tests/dialects/test_transform.py [hunk 10]
```diff
@@ -76,10 +75,198 @@ def test_tileop_init():
...
+def test_get_parent_op():
+    target = test.TestOp(result_types=[transform.AnyOpType()]).results[0]
+    op = transform.GetParentOp(target=target)
+    expected = """
+%0 = "transform.get_parent_op"(%1) <{"nth_parent" = 1 : i64}> : (!transform.any_op) -> !transform.any_op    """
+    assert_print_op(
+        op,
+        expected,
+        None,
+    )
```
Comment (superlopuh) [inline tests/dialects/test_transform.py:117]:
```suggestion
    expected = """
%0 = "transform.get_parent_op"(%1) <{"nth_parent" = 1 : i64}> : (!transform.any_op) -> !transform.any_op    """
    assert_print_op(
        transform.GetParentOp(target=target),
        """
%0 = "transform.get_parent_op"(%1) <{"nth_parent" = 1 : i64}> : (!transform.any_op) -> !transform.any_op    """,
        None,
    )
```
Probably worth doing the same throughout the file, will make it a bit easier to read

File: xdsl/dialects/transform.py (hunk 6 above)
Comment (JosseVanDelm) [reply xdsl/dialects/transform.py:569]: oh okay. Weird haha. Thanks for checking!

File: tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir (hunk 7 above)
Comment (JosseVanDelm) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir:1]: Let's remove it then!

File: mytest.mlir
Comment (superlopuh) [inline mytest.mlir:1]: seems like this got checked in accidentally

**Comment by superlopuh** [comment]
This is overall looking good, there are just small changes remaining. Could I just ask you to please make smaller pull requests in the future? A file that needs to be expanded explicitly in the GitHub UI is usually a signal to split up the changes into smaller diffs. For example, for this we could easily have merged the new attributes first, then the new operations.

File: mytest.mlir
Comment (JonasCrols) [reply mytest.mlir:1]: sorry, didn't notice! my bad...

**Review by jorendumoulin — APPROVED** [review]

File: tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir (hunk 2 above)
Comment (JonasCrols) [reply tests/filecheck/mlir-conversion/with-mlir/dialects/transform/transform_types.mlir:7]: I don't really see where custom syntax can be used in this case. are there any particular operations you're referring to?
